segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/WriterState.java

```diff
@@ -122,6 +129,57 @@ long getLastReadSequenceNumber() {
     void setLastReadSequenceNumber(long value) {
         Preconditions.checkArgument(value >= this.lastReadSequenceNumber.get(), "New LastReadSequenceNumber cannot be smaller than the previous one.");
         this.lastReadSequenceNumber.set(value);
+        recordReadComplete();
+    }
+
+    /**
+     * Indicates the fact that the {@link StorageWriter} has completed reading.
+     */
+    void recordReadComplete() {
+        val ffc = this.forceFlushContext.get();
+        if (ffc != null) {
+            ffc.setLastReadSequenceNumber(this.lastReadSequenceNumber.get());
+        }
+    }
+
+    /**
+     * Indicates the fact that the {@link StorageWriter} has completed a flush stage.
+     *
+     * @param result The {@link WriterFlushResult} summarizing the flush stage.
+     */
+    void recordFlushComplete(WriterFlushResult result) {
+        val ffc = this.forceFlushContext.get();
+        if (ffc != null && ffc.flushComplete(result)) {
+            this.forceFlushContext.set(null);
+            ffc.getCompletion().complete(ffc.isAnythingFlushed());
```

Comment:
What happens when there is nothing to flush?

Reply:
The StorageWriter may execute multiple iterations until it is able to complete this. When requested to force flush, it sets a context, and as long as that context is present, the Storage Writer will forcefully try to flush everything, read more data, flush, etc., until the requested "flush-to" sequence number has been processed.